### ISSUE TITLE
log10 appears broken in Numexpr 2.4.6 (also in 2.5)

### ISSUE THREAD

**Issue by 153957** [pydata/numexpr#210]
An installation of PyTables (v3.2.2) using version 2.4.6 of Numexpr fails queries which use log10.
The details are reported in a PyTables issue: https://github.com/PyTables/PyTables/issues/534

I tried to update to Numexpr 2.5 but got the same results.

**Comment by FrancescAlted** [comment]
Ok.  If the problem is in numexpr it would help if you can send a self-contained example showing the problem.  Also, in case you find a fix, filing a pull request would be great.

**Comment by 153957** [comment]
Here is a small version of the data file we use.
It contains multiple values in the `energy` column.

Use the following script:
```python
import tables
import numpy
import numexpr
print numexpr.__version__
# 2.5
data = tables.open_file('corsika_overview_local.h5')
set(numpy.log10(data.root.simulations.col('energy')))
# {12.0, 13.0, 14.0, 15.0, 16.0, 17.0}
set(numpy.log10(data.root.simulations.read_where('log10(energy) < 14')['energy']))
# {12.0, 13.0, 14.0, 15.0, 16.0, 17.0}
set(numpy.log10(data.root.simulations.read_where('log10(energy) == 13')['energy']))
# {13.0, 15.0}
```
[corsika_overview_local.h5.zip](https://github.com/pydata/numexpr/files/193092/corsika_overview_local.h5.zip)

Unfortunately I do not have time to work on a fix nor to familiarize myself with the inner workings of numexpr.

**Comment by FrancescAlted** [comment]
Ok, thanks.  However, the example that you provided is not self-contained (it depends on PyTables and external HDF5 datasets).  Would you mind providing something that only depends on numexpr and data that is generated in the same script?

**Comment by 153957** [comment]
I am also not familiar enough with the inner workings of PyTables to know how it performs its query using numexpr such that I can make a more self-contained version. I believe you know more about that than I do. The only difference in the example above was a different version of numexpr (`conda install numexpr==x.y.z`).

**Comment by tomkooij** [comment]
Unfortunately this seems to be a problem in the conda package starting from 2.4.6.
This bug does not appear in a locally installed and compiled numexpr.

**Comment by FrancescAlted** [comment]
Great to know.  Thanks @tomkooij for reporting.  Will close this for now.

**Comment by tomkooij** [comment]
It looked like a conda package problem, but it turns out conda packages started using Intel MKL starting from numexpr 2.4.6.

pytables ```table.read_where()``` is broken when using functions (such as ```log10()```) in combination with numexpr/MKL.

This is a testcase without depending on pytables. This simulates a read_where() query using a numpy.rec.array and compares this with a query on a single column.
 
https://gist.github.com/tomkooij/461de490f60e290d201d9c8eefc6c90f

From debugging the numexpr VM it looks like VEC_ARG1_VML is to blame: https://github.com/pydata/numexpr/blob/master/numexpr/interp_body.cpp#L84
The vectorizes VML function should step over the other columns in memory. VEC_ARG1 seems to do this but VEC_ARG1_VML does not. Unfortunately I cannot debug this using gdb.

**Comment by FrancescAlted** [comment]
With your self-contained script it has been pretty easy to spot the problem.  I  have just found a fix for that.

**Comment by tomkooij** [comment]
@FrancescAlted: Thanks for the quick fix. **Unfortunately, the code above is still broken ** And thus pytables read_where() queries using VML functions is still broken.

9169600 fixes ```numexpr.evaluate('log10(...)'``` but the script above (and pytables.read_where() use ```NumExpr.run()```. The patched code in 9169600 is then bypassed and a VML function is run on non-contiguous memory.

This affects all queries using functions in pytables.read_where() when VML is used. I tested with sqrt() and log10() and both are broken. 

New self-contained test-script, based on script above:
```
"""
Simulate a read_where() query on a pytables table
"""
import numpy
import numexpr
from numexpr.necompiler import NumExpr, stringToExpression


def make_NumExpr(expr):
    """ Make a NumExpr object exactly as pytables """

    mysig = [('f1', float)]
    typemap = {'f1': float}
    strexpr = stringToExpression(expr, typemap, {})
    print "DEBUG: strexpr ", strexpr

    return NumExpr(strexpr, mysig)

def make_recarray():
    """ make a pytables table-like numpy.rec.array """

    data = numpy.rec.array(None, formats=('f4,f4,f4'), shape=(5,))
    data['f0']= numpy.arange(5)
    data['f1']= numpy.arange(200,205)
    data['f2']= numpy.arange(5)

    return data

if __name__ == '__main__':
    print "numexpr version: ", numexpr.__version__
    print "VML: ", numexpr.get_vml_version()

    expr_log10 = make_NumExpr('log10(f1)')
    expr_sqrt = make_NumExpr('sqrt(f1)')
    table = make_recarray()

    print "log10() = (2.0x, 2.0x, ..) ", expr_log10(table['f1'])
    print "sqrt() = (14.1, 14.1, ...)", expr_sqrt(table['f1'])

```


Output of the script above:
```
numexpr version:  2.5.2.dev0
VML:  Intel(R) Math Kernel Library Version 11.3.1 Product Build 20151021 for Intel(R) 64 architecture applications
DEBUG: strexpr  FuncNode(log10, float, (VariableNode(f1, float, ()),))
DEBUG: strexpr  FuncNode(sqrt, float, (VariableNode(f1, float, ()),))
log10() = (2.0x, 2.0x, ..)  [ 2.30102992        -inf  0.          2.30319595  0.        ]
sqrt() = (14.1, 14.1, ...) [ 14.14213562   0.           1.          14.17744732   1.        ]

```
The output should be a continuous row of number, but it is easy to see that the other columns are intermixed. A VML function is run on non-contiguous memory.

**Comment by tomkooij** [comment]
This seems to be a pytables problem, not a numexpr bug.

In pytables.conditions ```getContext()``` is not called before calling ```NumExpr()```:

https://github.com/PyTables/PyTables/blob/31abd13bdd587c39123849c76a67aacac1f95066/tables/conditions.py#L405

I will report it at pytables as well.

**Comment by FrancescAlted** [comment]
Umph.  Yeah, PyTables uses non-public API for accessing numexpr.  It would be nice to change that, but unfortunately I don't that will happen any time soon.  Meanwhile, could you reopen the ticket in PyTables?  Also, could you try to modify the PyTables code and pass `ex_uses_vml=True` in the NumExpr() call above and see if that works?  Thanks.

**Comment by tomkooij** [comment]
Passing ```èx_uses_vml=True``` seems to fix it.

I will reopen the PyTables issue and make a PR.

**Comment by tomkooij** [comment]
While working on Pytables I discovered 'abs()' is broken. Fixed in #212.
